### PR TITLE
native promises

### PR DESCRIPTION
**Pull request by rogierschouten**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -6,7 +6,6 @@
   "scripts": {
     "test": "npm run lint && npm run build && mocha dist/test/test-*.js",
     "cover": "npm run build && istanbul cover _mocha -- dist/test/test-*.js",
-    "prepublish": "npm run test",
```
Comment (poelstra) [inline package.json]: I use this such that it actually builds and tests, when I type `npm publish`.
I'd like to keep this behavior intact.

File: package.json [hunk 2]
```diff
@@ -5,7 +5,7 @@
   "main": "./dist/lib/index.js",
   "scripts": {
     "test": "npm run lint && npm run build && mocha dist/test/test-*.js",
-    "cover": "npm run build && istanbul cover _mocha -- dist/test/test-*.js",
+    "cover": "npm run build && istanbul cover ./node_modules/mocha/bin/_mocha -- dist/test/test-*.js",
```
Comment (poelstra) [inline package.json]: Unnecessary: npm already adds this to PATH.

File: src/examples/hello-world-backpressure.ts [hunk 3]
```diff
@@ -6,11 +6,11 @@
...
  */
 
-import Promise from "ts-promise";
 import Stream from "../lib/index";
+import {delay} from "../lib/util";
 
 Stream.from([1, 2, 3, 4])
-	.map((n) => Promise.resolve(n * 2).delay(1000))
-	.forEach((n) => console.log(n));
+	.map((n: number) => Promise.resolve(n * 2).then((k: number) => delay(1000, k)))
```
Comment (poelstra) [inline src/examples/hello-world-backpressure.ts]: Please leave out the TS-specific things, as these examples are intended to also be copy-pasteable to e.g. README and should be understandable as 'plain' ES6.

File: src/lib/Stream.ts [hunk 4]
```diff
@@ -8,11 +8,11 @@
 
 "use strict"; /* istanbul ignore next */ // ignores Typescript's __extend() function
 
-import { Promise, Thenable, BaseError, Deferred } from "ts-promise";
 import * as assert from "assert";
 
+import BaseError from "./BaseError";
 import { Transform, map, filter } from "./Transform";
-import { swallowErrors } from "./util";
+import {defer, Deferred, swallowErrors, track, TrackedPromise} from "./util";
```
Comment (poelstra) [inline src/lib/Stream.ts]: Please leave the spaces around import statements, to be consistent with the rest.

File: package.json (hunk 2 above)
Comment (rogierschouten) [reply package.json]: it doesn't work for me unless I add that path

File: package.json (hunk 1 above)
Comment (rogierschouten) [reply package.json]: ok

File: src/examples/hello-world-backpressure.ts (hunk 3 above)
Comment (rogierschouten) [reply src/examples/hello-world-backpressure.ts]: ok

File: src/lib/Stream.ts (hunk 4 above)
Comment (rogierschouten) [reply src/lib/Stream.ts]: ok

**Comment by rogierschouten** [comment]
- es5 emit did not change any javascript code w.r.t. es2015, so that's fine
- I can't explain the coverage decrease as the methods that are un-covered were previously un-covered as well. Perhaps the size of the javascript output has changed

File: package.json (hunk 2 above)
Comment (poelstra) [reply package.json]: Hmmm, you're right: apparently this indeed doesn't work on Windows: https://www.npmjs.com/package/istanbul#usage-on-windows

So, I'm fine with leaving it like you proposed.

**Comment by poelstra** [comment]
Thanks! Looks great. Changes to BaseError look tricky, but I'm convinced by Melvin's test results that they Should Work :tm:

**Comment by poelstra** [comment]
I'll run some further tests on this at home, and if they pass, will bump the version and publish.

**Pull request merged**